File: server/pipeline.py
"""Reusable voice turn: STT -> agent -> TTS, framework-agnostic."""

import asyncio
import logging
import time

logger = logging.getLogger("voice.pipeline")


async def run_voice_turn(stt, agent, tts, audio: bytes, history: list[dict], *, send, get_cards=None) -> None:
    """Run one push-to-talk turn and report progress through `send(kind, payload)`.

    `send` receives:
      - ("debug", {"event": str, "data": dict})
      - ("transcript", {"text": str})
      - ("assistant_text", {"text": str})
      - ("entities", {"entities": list})
      - ("error", {"message": str})
      - ("wav", bytes)

    `history` is mutated in place and rolled back if the agent fails.
    """
    t0 = time.perf_counter()
    text = await asyncio.to_thread(stt.transcribe, audio)
    await send(
        "debug",
        {"event": "stt", "data": {"text": text, "latency_ms": round((time.perf_counter() - t0) * 1000)}},
    )
    if not text:
        await send("error", {"message": "I didn't catch that"})
        return
    await send("transcript", {"text": text})

    touched: list[str] = []

    async def on_agent_event(event: str, data: dict) -> None:
        if event == "touched":
            for eid in data.get("entity_ids", []):
                if eid not in touched:
                    touched.append(eid)
            return
        await send("debug", {"event": event, "data": data})

    checkpoint = len(history)
    try:
        reply = await agent.run(history, text, on_event=on_agent_event)
    except Exception:
        logger.exception("agent failure")
        del history[checkpoint:]
        await send("error", {"message": "LLM backend offline"})
        return

    await send("assistant_text", {"text": reply})
    if get_cards is not None and touched:
        try:
            cards = await get_cards(touched[:8])
            await send("entities", {"entities": cards})
        except Exception:
            logger.exception("entities refresh failed")

    t1 = time.perf_counter()
    wav = await asyncio.to_thread(tts.synthesize, reply)
    await send(
        "debug",
        {"event": "tts", "data": {"latency_ms": round((time.perf_counter() - t1) * 1000), "bytes": len(wav)}},
    )
    await send("wav", wav)

File: ts-backend/src/pipeline.ts
import type { TemperatureReading } from "./temperature.js";
import type { Agent, HistoryMessage } from "./agent.js";
import { InferenceError, type InferenceClient } from "./inference.js";

/** The sidecar's own explanation when it gave one, else a generic message. */
function speechError(e: unknown): string {
  return e instanceof InferenceError && e.detail ? e.detail : "speech engine offline";
}

export type SendFn = (kind: string, payload: any) => void | Promise<void>;

/** Run one push-to-talk turn: STT -> agent -> TTS. Transport-agnostic. */
export async function runVoiceTurn(
  inference: InferenceClient,
  agent: Agent,
  audio: Buffer,
  history: HistoryMessage[],
  send: SendFn,
  voice: string,
  getCards?: (ids: string[]) => Promise<any[]>,
): Promise<void> {
  const t0 = performance.now();
  let text: string;
  try {
    text = await inference.transcribe(audio);
  } catch (e) {
    await send("error", { message: speechError(e) });
    return;
  }
  await send("debug", {
    event: "stt",
    data: { text, latency_ms: Math.round(performance.now() - t0) },
  });
  if (!text) {
    await send("error", { message: "I didn't catch that" });
    return;
  }
  await send("transcript", { text });

  const touched: string[] = [];
  let reading: TemperatureReading | null = null;
  let expression: Record<string, unknown> | null = null;
  let activity: { tool: string; domain: string; service?: string } | null = null;
  const onAgentEvent = async (event: string, data: any) => {
    if (event === "reading") {
      reading = data;
      return;
    }
    if (event === "expression") {
      expression = data;
      return;
    }
    if (event === "touched") {
      for (const eid of data.entity_ids ?? []) {
        if (!touched.includes(eid)) touched.push(eid);
      }
      return;
    }
    // Product activity is separate from diagnostic text, including untruncated
    // success/failure information. Tool calls within a voice turn are sequential.
    if (event === "tool_call") {
      activity = null;
      const args = data.args;
      if ((data.name === "call_service" || data.name === "get_entities") &&
          args && typeof args === "object" && typeof args.domain === "string") {
        activity = {
          tool: data.name,
          domain: args.domain,
          ...(typeof args.service === "string" ? { service: args.service } : {}),
        };
        await send("activity", { ...activity, phase: "start" });
      }
    } else if (event === "tool_result" && activity?.tool === data.name) {
      await send("activity", { ...activity, phase: data.ok ? "complete" : "error" });
      activity = null;
    }
    await send("debug", { event, data });
  };

  const checkpoint = history.length;
  let reply: string;
  try {
    reply = await agent.run(history, text, onAgentEvent, { expressions: true });
  } catch {
    history.splice(checkpoint);
    await send("error", { message: "LLM backend offline" });
    return;
  }

  await send("assistant_text", { text: reply });
  if (getCards && touched.length) {
    try {
      await send("entities", { entities: await getCards(touched.slice(0, 8)) });
    } catch {
      /* card refresh is best-effort */
    }
  }

  const t1 = performance.now();
  let wav: Buffer;
  try {
    wav = await inference.synthesize(reply, voice);
  } catch (e) {
    await send("error", { message: speechError(e) });
    return;
  }
  await send("debug", {
    event: "tts",
    data: { latency_ms: Math.round(performance.now() - t1), bytes: wav.length },
  });
  // Begin the expression with playback, so slow synthesis cannot use up its lifetime.
  if (expression) await send("expression", expression);
  if (reading) await send("reading", reading);
  await send("wav", wav);
}
